### DISCUSSION TITLE
Set value to Options picker not working

### DISCUSSION THREAD

**Discussion by juangbm** [Budibase/budibase#9449] [Help] [answered]
Hi,

I have an Options Picker linked to a DataProvider to display some options, but I need to select as a default value the option that is stored in the database, And it is impossible, I tried everything, but no way to make this work. 
![image](https://user-images.githubusercontent.com/22911293/214808381-2c8fd0ef-43ba-4b37-b8fb-71be69c7753a.png)
I tried using _id, as it is explained here https://github.com/Budibase/budibase/discussions/4579 , but I got a null as Default value, I also tried using the ID from the database (used as value when setting the Options Picker), also assigning the text displayed as Default Value, and nothing works ...
Can anybody tell mewhat can be going on?

**Comment by melohagan** [comment]
Hey @juangbm 

What does the options picker source look like? https://docs.budibase.com/docs/option-picker#displaying-options-in-the-options-picker

**Reply by juangbm** [reply]
Hi @melohagan 

The source I am using is a Data Provider:
![image](https://user-images.githubusercontent.com/22911293/214880885-b73bf44f-7f9d-4c72-b0aa-b8e446d9c938.png)

The Data provider looks like:
![image](https://user-images.githubusercontent.com/22911293/214881345-59d4f1b8-6eb1-4579-8d9d-66b6ea7ff916.png)

Thanks!

**Reply by melohagan** [reply]
Thanks, and what data-type is *idCLU_MERCHANT*, and which value were you trying to use as the default?

**Reply by juangbm** [reply]
idCLU_MERCHANT is an Number

![image](https://user-images.githubusercontent.com/22911293/214886675-562cfe11-4d43-437f-93e8-d3e86f080ad6.png)

I tied almost whith everything... including static values (1,2) of items that were in the options picker, I also tried using the _id, getting it  from the DetailRepeater (at the top of the Form...)

**Reply by melohagan** [reply]
Ah, I think I know what this is. Handlebars always return a string. So `1 !== "1"`.

You may need to use a JavaScript binding to get the default value to be appropriate, i.e. `return 1`

**Reply by juangbm** [reply]
Thanks @melohagan ! It worked!

However, now I have a problem saving that in the DB ... :( 
I am using this OptionsPicker to display a relationship (filtering values that should appear there) and after selecting on, I have to save the changes... and I am getting the error: **row[key].map is not a function**
The name of the Field from the options picker is PARENT_SELECTED, so what I do when saving is:

![image](https://user-images.githubusercontent.com/22911293/214908180-ca3b8a0a-9429-4dd4-b171-327b23fa943b.png)

where CLU_MERCHANT_PARENT is the field in the database defined in the relationship to save:

![image](https://user-images.githubusercontent.com/22911293/214908534-7b051119-f400-41a1-ab75-53624f352b1f.png)

Any idea about how can I do it?

**Reply by melohagan** [reply]
Hey @juangbm 

I believe the issue is that the data-type for *PARENT_SELECTED* is a **number**, whereas the data-type of CLU_MERCHANT_PARENT is a **link** (relationship).

Ordinarily the relationship picker should be used instead of the options picker for relationships, however I appreciate you can't because you need to filter the available relationship options. This will be fixed in future.

In the meantime, you may be able to get this to work by encoding the number. So using a JS binding like so:

```js
const num = $("FormDetail.Fields.PARENT_SELECTED");
return helpers.encodeURI(`['${num}']`);
```

**Reply by juangbm** [reply]
Hi @melohagan 

I tried your solution adding your code on saving action, but I still get the same result row[key].map is not a function

![image](https://user-images.githubusercontent.com/22911293/215083788-a081044d-57e1-4c87-841c-d617512cfb6f.png)

Any idea?

Thanks very much!!

**Reply by melohagan** [reply]
Hmm, I may need to see the app export, if you are able to send that through to support@budibase.com ?

**Reply by juangbm** [reply]
Sent, thanks!

**Comment by melohagan** [comment, marked as answer]
Thank you for the app export @juangbm 

I was able to determine that this is a bug, and I picked up the issue you raised here: https://github.com/Budibase/budibase/issues/9375

Hopefully the fix will be out for the upcoming release in a week's time.

**Reply by peterhuppertz** [reply]
On previous versions of Budibase (< 2.2) I could set the default value of a relationship picker like outlined in the discussion here:
```
const id = $("db.Rows.0.fk_users");
return helpers.encodeURI(`['${id}']`);
```

However, this stopped working in the latest release of budibase.

Weird enough it still works when I hardcode the value as default, in that case: `%5B'us_c81da1f7f2f04e70bcb46cbd67ea44d0'%5D`. That value get selected as default value. As soon as the variable is involved it stops working and shows `null` as default value, although their output look EXACTLY the same. Any ideas how to get it work again?

**Comment by bjornclauw** [comment]
Hi, I resolved this the other way around. 

in my option picker i set the _id field in the value column
![image](https://user-images.githubusercontent.com/32811735/215586235-5b37820e-494d-4c5d-be04-eceb504b069f.png)

So now I can use this value when I'm saving the row
![image](https://user-images.githubusercontent.com/32811735/215586413-48b34c69-a3cd-4d0f-a528-839fcd1030ee.png)

Now for the live filtering of my cascading option  picker(s) (each one has his own data provider)
I use following filter:

![image](https://user-images.githubusercontent.com/32811735/215586890-28e90330-6c47-449b-a5fc-0ebfd2a770bd.png)


```js
return $("edit Form.Fields.edit option picker 1")[0].slice(3,-3);
// %5B7%5D --> 7
```

**Reply by bjornclauw** [reply]
to clarify:
pos1_fkey is a real column in my postgres table, the pos1 is the field generated by the relationship.

**Reply by bjornclauw** [reply]
And for default value for the options picker -> {{ Repeater.assets.pos1.[0]._id }}

**Reply by melohagan** [reply]
Good job for finding the workaround yourself!

**Comment by bjornclauw** [comment]
@melohagan I was wondering if it would be possible to add a filter button to a relationship picker (like we use on the parent data provider for our options picker). If I'm not mistaken, this would remove the need to use javascript entirely for this usecase.

**Reply by melohagan** [reply]
Hey @bjornclauw We definitely do intend to do that. I'd started it, but it needs a bit of a refactor: https://github.com/Budibase/budibase/pull/8897

**Comment by apparentlyjeff** [comment]
I was also having the issue where setting the default value on the options picker was resulting in it displaying, "null" in the form.  I tried implementing the storage of _id instead of my integer id field, and it worked, but wasn't ideal.  After playing with it for far too long, I discovered that in the "Default Value", instead of simply placing {{ DATA PROVIDER.ROWS.0.TYPE_ID }} in the handlebars section, that if I go to javascript and simply do this, it works: 'return $("DATA PROVIDER.ROWS.0.TYPE_ID");'

**Reply by chadgriff** [reply]
@apparentlyjeff Thanks for your post...this was driving me crazy and your simple approach using javascript and it worked for me as well.

**Reply by melohagan** [reply]
Handlebars -> always returns text, unless you use the literal helper, e.g. `{{ literal Data Provider.Rows.0.TYPE_ID }}`

JavaScript -> returns the object keeping the type

**Comment by hawkinsjon** [comment]
> I was also having the issue where setting the default value on the options picker was resulting in it displaying, "null" in the form.  I tried implementing the storage of _id instead of my integer id field, and it worked, but wasn't ideal.  After playing with it for far too long, I discovered that in the "Default Value", instead of simply placing {{ DATA PROVIDER.ROWS.0.TYPE_ID }} in the handlebars section, that if I go to javascript and simply do this, it works: 'return $("DATA PROVIDER.ROWS.0.TYPE_ID");'  

Thank you for this! Solved my problem.
